### PR TITLE
Content Recommend block setting update

### LINKED ISSUE #379: Recommended Content settings iteration
### Is your enhancement related to a problem? Please describe.

The current Recommended Content block takes input from a user via several controls, two of which are the Categories and Tags multiselect boxes.  Let's look into an iteration here that:

1. Sets the `Post type` field to the same post type for the currently edited post type (e.g., if editing on a Page, then set the `Post type` to `Page`).
2. Set the `Categories` and `Tags` to be the same as the currently edited post type (e.g., if editing on a Page that's set as `Categories`=`Sports`, `Tags`=`[Soccer, Technology]`, then set those categories and tags as the default ones for the Recommended Content block).
3. Add a `Use this Post's Categories and Tags` toggle that defaults to enabled and sets those fields based on #2 above.  This makes those fields uneditable (or alternatively hides them).  Disabling this toggle makes the `Categories` and `Tags` fields editable.

Props to @jakemgold for the inspiration on this UX improvement.

### Designs

Use This Posts Categories And Tags setting enabled and showing full view of pre-set fields:
UseThisPostsCategoriesAndTags_EnabledFullView

Use This Posts Categories And Tags setting enabled but hiding the pre-set fields / collapsing them away from view:
UseThisPostsCategoriesAndTags_EnabledCollapsed

UseThisPostsCategoriesAndTags_Disabled:
UseThisPostsCategoriesAndTags_Disabled

### Describe alternatives you've considered

n/a

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by jayedul**
### Description of the Change
Content Recommend block setting updated

Closes #379 

### Changelog Entry

> Changed - Post type and taxonomy selection enhanced

### Credits

Props @jayedul, @jakemgold 

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests pass.

### CODE REVIEW

**Comment by peterwilsoncc** [comment]
I did a quick test of this and think it will need to follow the guide for [block deprecation](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/) in the block editor handbook.

When the attributes change/the block is altered the deprecation code needs to be added to safely migrate from one form to another.

At the moment, blocks created on the default branch (develop) fail on this new branch -- I presume because the attributes and/or saved data are not in the same format.

testing-branch

**Comment by peterwilsoncc** [comment]
I've also noticed an issue with the new format on the front end:

1. On `develop` create a recommended content block and publish the post (ensure it has results)
2. Visit the post on the front end, the post results display as expected.
   Screen Shot 2023-02-23 at 10 46 47 am
3. Switch to the feature branch for this PR
4. Run `npm run build`
5. No results are displayed
   Screen Shot 2023-02-23 at 10 48 23 am

**Comment by jayedul** [comment]
Post not found issue is fixed. However I couldn't reproduce block crashing issue. Here's a video showing it works here after changing develop to fix branch. Which version of WordPress did you test with? @peterwilsoncc 

https://user-images.githubusercontent.com/12506700/224363249-47d064b8-51cd-49ae-97b5-bcf1805edb36.mp4

**Comment by jayedul** [comment]
BTW, I guess e2e test is failing here because of cypress-wp-utils library. 
Most likely this fix will resolve it. https://github.com/10up/cypress-wp-utils/pull/89/files#diff-a807a69cfe758f8aabc6ea3a6f836df2dceaff610de1bd4384a6ed0afca4dc63R79

**Comment by jeffpaul** [comment]
I'll defer to Peter's re-review here, though once that's approved I will want to do a live test on this PR to validate it updates the settings as expected before we merge this in

**Comment by peterwilsoncc** [comment]
@jayedul I think I was testing against master.

I'll document the full steps. Make sure to run the build step each time you switch branches.

1. checkout `develop`
2. `npm run build`
3. Create a new post
4. Add a recommend content block
5. In the block settings add some tags to relate content by
   Screen Shot 2023-03-14 at 11 09 23 am
6. Save Post
7. Checkout the feature branch
8. `npm run build`
9. Reload the edit post interface
10. The block crashes due to the modifications 

I suspect it's due to the change from `{"taxQuery":{"post_tag":[3,17,18]}}` to `{"contentPostType":"post","taxQuery":{"post_tag":[4]},"defaultTermsSetFor":["post_tag","post_tag","post_tag"]}` in the attribute.

**Comment by jayedul** [comment]
It's made ready for functional review, though eslint test is failing. Locally fixed code by eslint command fails here, which is contradictory.

**Comment by jayedul** [comment]
This ticket can be re-reviewed.

File: includes/Classifai/Blocks/recommended-content-block/edit.js
Comment (dkotter) [inline includes/Classifai/Blocks/recommended-content-block/edit.js:126]: Kind of confused on why we're making a custom ajax request here using jQuery. Can't we use something like [getEntityRecords](https://developer.wordpress.org/block-editor/reference-guides/data/data-core/#getentityrecords) to get taxonomy information?

File: includes/Classifai/Blocks/recommended-content-block/block.json
Comment (dkotter) [inline includes/Classifai/Blocks/recommended-content-block/block.json:22]: If I read the issue correctly, it sounds like we want this enabled by default, not turned off by default

File: includes/Classifai/Blocks/recommended-content-block/inspector-controls/taxonomy-controls.js
Comment (dkotter) [inline includes/Classifai/Blocks/recommended-content-block/inspector-controls/taxonomy-controls.js]: I know the original issue mentioned just categories and tags but I think we allow setting any taxonomies that are assigned to the current post type. Ideal would be to not hardcode these values but support all taxonomies the current post type supports

File: includes/Classifai/Blocks/recommended-content-block/inspector-controls/taxonomy-controls.js
Comment (dkotter) [inline includes/Classifai/Blocks/recommended-content-block/inspector-controls/taxonomy-controls.js]: Ideally we don't hardcode the word `Post` here, as this could be another type of content (page or CPT)

File: includes/Classifai/Blocks/recommended-content-block/markup.php
Comment (dkotter) [inline includes/Classifai/Blocks/recommended-content-block/markup.php:34]: This looks to be a new filter. We should add a filter docblock around this

File: includes/Classifai/Blocks/recommended-content-block/register.php
Comment (dkotter) [inline includes/Classifai/Blocks/recommended-content-block/register.php]:
```suggestion
		wp_send_json_error( array( 'message' => 'Invalid nonce' ) );
```

Comment (dkotter) [reply includes/Classifai/Blocks/recommended-content-block/register.php]: Probably should wrap this in a translation function as well

File: includes/Classifai/Blocks/recommended-content-block/register.php
Comment (dkotter) [inline includes/Classifai/Blocks/recommended-content-block/register.php]: I think we want to use `wp_unslash` on all these `$_GET` variables since WordPress may add slashes (see https://github.com/WordPress/wordpress-develop/blob/c421c7ead6c72ebb448c7a21f6bef3acfaab6a92/src/wp-includes/load.php#L1123-L1132)

Comment (dkotter) [reply includes/Classifai/Blocks/recommended-content-block/register.php]: Probably also want to verify that this post ID actually belongs to a post object

File: includes/Classifai/Blocks/recommended-content-block/register.php
Comment (dkotter) [inline includes/Classifai/Blocks/recommended-content-block/register.php]:
```suggestion
	$attributes = $blocks[ $index ] ?? : null;
```

File: includes/Classifai/Blocks/recommended-content-block/register.php
Comment (dkotter) [inline includes/Classifai/Blocks/recommended-content-block/register.php]: I think we can invert the conditions here and return `$attributes` early if the conditions don't match, rather than wrap all this code in a conditional

**Review by dkotter — CHANGES_REQUESTED** [review]

File: includes/Classifai/Blocks/recommended-content-block/block.json
Comment (jayedul) [reply includes/Classifai/Blocks/recommended-content-block/block.json:22]: This value controls the switch. By default currently associated terms will be selected automatically on `Content Recommend` block mount. Then user can modify the term selection. However if they enable switch (usePostTerms technically), then term input field will disappear and current post terms will be used in filtering no matter what is in input field. 

Screenshot 2023-04-19 at 7 46 52 AM

**Comment by jeffpaul** [comment]
@dkotter @peterwilsoncc back to you for re-review

File: includes/Classifai/Blocks/recommended-content-block/block.json
Comment (dkotter) [reply includes/Classifai/Blocks/recommended-content-block/block.json:22]: So this isn't currently working for me, unless I'm doing something wrong. I have a post that has 2 categories and 1 tag set but when I add the Recommended Content block, neither of those inputs are filled out.

That said, I still think the desired behavior is the `Use assigned terms` should be selected by default. Then if you deselect that, we can either leave the taxonomy options empty or at that point, maybe we set those to match the current values of the post? Any thoughts on how you expect this to work @jeffpaul?

Comment (dkotter) [reply includes/Classifai/Blocks/recommended-content-block/block.json:22]: @jayedul Talked to Jeff on this and the direction we came to is that the `Use assigned terms` should be selected by default, which is opposite of what we have now. And then if someone deselects that option, we should then pre-fill in whatever terms the post has set (this doesn't seem to be working right now)

**Comment by jeffpaul** [comment]
@10up/open-source-practice pinging the team in hopes of getting this through another round of code review/testing feedback

**Comment by iamdharmesh** [comment]
@jeffpaul It seems there are some [open](https://github.com/10up/classifai/pull/397#discussion_r1165972006) feedback that need to be addressed. Also, we have to manage cases like using recommended block outside of edit posts/pages. For example, adding a block in the footer template, where we don't have terms assigned to auto-select as described [here](https://github.com/10up/classifai/pull/397#discussion_r1195383709). I will try to update PR to address these issues.

**Comment by jeffpaul** [comment]
For cases where there aren't terms to auto-select then the terms should be empty.

**Comment by jeffpaul** [comment]
@ravinderk as you're able during your OSS time, could you please look to handle the code review feedback and any open tasks on this PR to help move it along for final review/merge?  Thanks!

**Comment by dkotter** [comment]
Due to [Personalizer going away](https://github.com/10up/classifai/issues/392#issuecomment-1759809106), thinking we probably just want to close this out.

**Comment by jeffpaul** [comment]
:sad-panda:  I agree.

**Pull request closed without merging**
